sihl/src/web_rest.ml

```diff
@@ -0,0 +1,352 @@
+    let thing = Conformist.decode_and_validate schema urlencoded in
+    let id = Opium.Router.param req "id" in
+    match thing with
+    | Ok thing ->
+      let* updated = Service.update id thing in
+      (match updated with
+      | Ok _ ->
+        Opium.Response.redirect_to (Format.sprintf "/%s/%s" name id)
+        |> Web_flash.set_notice
+             (Format.sprintf "Successfully updated %s" (singularize name))
+        |> Lwt.return
+      | Error msg ->
+        Opium.Response.redirect_to (Format.sprintf "/%s/%s/edit" name id)
+        |> Web_flash.set_alert msg
+        |> Form.set [] urlencoded
+        |> Lwt.return)
+    | Error errors ->
+      Opium.Response.redirect_to (Format.sprintf "/%s/%s/edit" name id)
+      |> Web_flash.set_alert "Some of the input is invalid"
+      |> Form.set errors urlencoded
+      |> Lwt.return
+  ;;
+
+  let delete' name req =
+    let open Lwt.Syntax in
+    let id = Opium.Router.param req "id" in
+    let* thing = Service.find id in
+    match thing with
+    | None ->
+      Opium.Response.redirect_to (Format.sprintf "/%s" name)
+      |> Web_flash.set_alert
+           (Format.sprintf
+              "%s with id '%s' not found"
+              (singularize (capitalize name))
+              id)
+      |> Lwt.return
+    | Some thing ->
+      let* result = Service.delete thing in
+      (match result with
+      | Ok () ->
+        Opium.Response.redirect_to (Format.sprintf "/%s" name)
+        |> Web_flash.set_notice
+             (Format.sprintf
+                "Successfully deleted %s '%s'"
+                (singularize name)
+                id)
+        |> Lwt.return
+      | Error msg ->
+        Opium.Response.redirect_to (Format.sprintf "/%s" name)
+        |> Web_flash.set_notice
+             (Format.sprintf "Failed to delete %s: '%s'" (singularize name) msg)
+        |> Lwt.return)
+  ;;
+end
+
+type action =
+  [ `Index
+  | `Create
+  | `New
+  | `Edit
+  | `Show
+  | `Update
+  | `Destroy
+  ]
+
+let router_of_action
+    (type a)
+    (module Controller : CONTROLLER with type t = a)
+    name
+    schema
+    (action : action)
+  =
+  match action with
+  | `Index -> Web.get (Format.sprintf "/%s" name) (Controller.index name)
+  | `Create ->
+    Web.post (Format.sprintf "/%s" name) (Controller.create name schema)
+  | `New -> Web.get (Format.sprintf "/%s/new" name) (Controller.new' name)
+  | `Edit -> Web.get (Format.sprintf "/%s/:id/edit" name) (Controller.edit name)
+  | `Show -> Web.get (Format.sprintf "/%s/:id" name) (Controller.show name)
+  | `Update ->
+    Web.put (Format.sprintf "/%s/:id" name) (Controller.update name schema)
+  | `Destroy ->
+    Web.delete (Format.sprintf "/%s/:id" name) (Controller.delete' name)
+;;
+
+let routers_of_actions
+    (type a)
+    name
+    schema
+    (module Controller : CONTROLLER with type t = a)
+    (actions : action list)
+  =
+  let rec create_routers (actions : action list) (routers : Web.router list) =
```

Comment:
Can this not be replaced with a `List.map (...) actions`?

Reply:
it can, fixed :)